Cython/TestUtils.py

```diff
@@ -228,21 +235,96 @@ def unpack_source_tree(tree_file, workdir, cython_root):
     return workdir, header
 
 
+def clear_function_and_Dependencies_caches():
+    Utils.clear_function_caches()
+    Dependencies._dep_tree = None  # discard method caches
+
+
+def fresh_cythonize(*args, **kwargs):
+    """Clear function caches and run cythonize"""
+    clear_function_and_Dependencies_caches()
+    Dependencies.cythonize(*args, **kwargs)
+
+
+def touch_file(path):
+    """Set the access and modified times of the file
+    specified by path as the current time"""
+
+    os.utime(path, None)
+
+
+def relative_lines(lines, line, start, end, fallback=None):
+    """Returns the lines specified by `start` and `end` relative to `line`.
+
+    If `fallback` (Sized Iterable) is specified, it will be used for `line`, `start`, `end`
+    applied to `lines` to generate a message for a ValueError
+    if the original `line` is not found in `lines`."""
+
+    if fallback:
+        if not isinstance(fallback, Sized):
+            raise TypeError("'fallback' must be a Sized instance")
+
+        if not isinstance(fallback, Iterable):
+            raise TypeError("'fallback' must be an instance of the Iterable")
+
+        if len(fallback) != 3:
+            raise ValueError("'fallback' must contain three values"
+                             " for 'line', 'start', 'end'")
+
+    try:
+        ind = lines.index(line)
+        return lines[ind+start: ind+end]
+    except ValueError as e:
+        if fallback is None:
+            raise e
+
+        fallback_lines = map(repr, relative_lines(lines, *fallback))
+
+        raise ValueError(
+            "{0!r} was not found, presumably in \n{1}".format(
+                line, "\n".join(fallback_lines)))
+
+
+def relative_lines_from_file(path, line, start, end, fallback=None):
+    """Same as `relative_lines`, but uses `path`
+    as source for `lines` and returns joined lines."""
```

Comment:
It seems to me that copying a few lines from `relative_lines` here is not the best approach.

Reply:
i think it works fine, i resolving this